src/katgpucbf/utils.py

```diff
@@ -53,3 +58,48 @@ def parse_source(value: str) -> list[tuple[str, int]] | str:
         return [(ep.host, ep.port) for ep in endpoints]
     except ValueError:
         return value
+
+
+class DeviceStatus(Enum):
+    """Discrete `device-status` readings.
+
+    Doesn't follow the convention of :class:`aiokatcp.Sensor.Status` because
+    it's useful for `UNKNOWN` to have the highest (or worst) value.
+    """
+
+    OK = 1
+    DEGRADED = 2
+    FAIL = 3
+    UNKNOWN = 4
+
+
+class DeviceStatusSensor(aiokatcp.AggregateSensor):
+    """Summary sensor for quickly ascertaining device status.
+
+    This takes its value from the worst status of its target set of sensors, so
+    it's quick to identify if there's something wrong, or if everything is good.
+    """
+
+    def __init__(self, target: aiokatcp.SensorSet):
+        super().__init__(target=target, sensor_type=DeviceStatus, name="device-status")
+
+    def update_aggregate(
+        self,
+        updated_sensor: Optional[aiokatcp.Sensor[_T]],
+        reading: Optional[aiokatcp.Reading[_T]],
+        old_reading: Optional[aiokatcp.Reading[_T]],
+    ) -> aiokatcp.Reading[DeviceStatus]:  # noqa: D102
+        # For device status it's far simpler just to re-calculate everything
+        # each time, than to try and maintain state.
+        worst_status: aiokatcp.Sensor.Status = aiokatcp.Sensor.Status.UNKNOWN
+        for sensor in self.target.values():
+            if self.filter_aggregate(sensor):
+                worst_status = max(worst_status, sensor.status)
+
+        if worst_status == aiokatcp.Sensor.Status.NOMINAL:
+            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.NOMINAL, DeviceStatus.OK)
```

Comment:
`sensor.timestamp` will be whichever sensor was last in `self.target` (it might even be self), which is probably not your intention. I didn't give as much thought to the timestamp as a I should have in the AggregateSensor design. What do you think of the following:
- If both `old_reading` and `reading` are not None (i.e. a sensor update), use `reading.timestamp`.
- Otherwise, use `time.time()`.

I'd also suggest always taking a max with the current `self.timestamp`, to avoid time warping if some sensors take time from the data stream and others from the host's clock.

Reply:
Yeah this is a good point. I hadn't thought through the timestamp very much, I'm implementing your approach.